### PR TITLE
Fix TopK agg exception due to partial type != final result     

### PR DESCRIPTION
**Pull request by matriv**
TopK uses a different partial structure (and therefore data type)
than the one used for the final result. As a consequence, in
`ProjectionBuilder#getAggregations` an agg signature is built, with
argument type a `TopKAggregation#StateType`, which if used down the
stuck in TopK aggregation, it will throw an error in the switch
statements which operate on the actual value type.

Therefore, check for that case in the ctor, and extract the inner type
of the `StateType`, to be used down the stack.

### CODE REVIEW

File: server/src/test/java/io/crate/integrationtests/AggregateExpressionIntegrationTest.java [hunk 1]
```diff
@@ -305,4 +308,33 @@ public void test_assure_cmp_by_function_call_with_reference_and_literal_does_not
...
         assertThat(response).hasRows("foo");
     }
+
+    @SuppressWarnings("unchecked")
+    public void test_topk_agg() {
+        execute("create table tbl (l long, l_no_doc_values long storage with(columnstore = false))");
+        execute("insert into tbl(l, l_no_doc_values) values (1, 1), (1, 1), (1, 2), (2, 2), (2, 2)");
+        execute("refresh table tbl");
+
+        // TODO: Check why the item and frequency numbers switch between Integer & Long
```
Comment (matriv) [inline server/src/test/java/io/crate/integrationtests/AggregateExpressionIntegrationTest.java:318]: I still haven't found why the resulting json (map) switches between Integer and Long for the value types of item and frequence. Will continue investigating and address it separately.

**Comment by mkleen** [comment]
The result value is determined from the bound signature from https://github.com/crate/crate/blob/f520aa3136013651954055a340a91e86fa5c2b05/server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java#L125

This argument should never be the internal state but always the type to which this function is bounded. So in this case here the bound signature has the argument type `TopKAggregation.State` but should be `Long`. We need to fix the signature binding.

File: server/src/test/java/io/crate/integrationtests/AggregateExpressionIntegrationTest.java [hunk 2]
```diff
@@ -305,4 +308,33 @@ public void test_assure_cmp_by_function_call_with_reference_and_literal_does_not
...
         assertThat(response).hasRows("foo");
     }
+
+    @SuppressWarnings("unchecked")
+    public void test_topk_agg() {
+        execute("create table tbl (l long, l_no_doc_values long storage with(columnstore = false))");
+        execute("insert into tbl(l, l_no_doc_values) values (1, 1), (1, 1), (1, 2), (2, 2), (2, 2)");
+        execute("refresh table tbl");
+
+        // Use this very verbose style of assertion, since the return type is an Array<Object(UNDEFINED)>,
```
Comment (matriv) [inline server/src/test/java/io/crate/integrationtests/AggregateExpressionIntegrationTest.java:318]: I could stick to `@UseJdbc(0)` and have more beautiful assertions, what do you think?

Comment (mfussenegger) [reply server/src/test/java/io/crate/integrationtests/AggregateExpressionIntegrationTest.java:318]: Either way is fine by me. I'd tend to keep the assertions as is and keep jdbc enabled

**Review by mfussenegger — APPROVED** [review]

**Comment by mfussenegger** [comment]
Maybe as follow up: Could we simulate this scenario in the aggregation test case to ensure we catch issues like this with the regular unit tests too?

**Comment by matriv** [comment]
> Maybe as follow up: Could we simulate this scenario in the aggregation test case to ensure we catch issues like this with the regular unit tests too?

good idea, I will try that as follow up, but independently, it's good to have at least one integ test, right?

**Comment by mfussenegger** [comment]
> > Maybe as follow up: Could we simulate this scenario in the aggregation test case to ensure we catch issues like this with the regular unit tests too?
> 
> good idea, I will try that as follow up, but independently, it's good to have at least one integ test, right?

Usually we don't create dedicated itests for scalars or aggregations (but docs typically contain an example)
Fine to keep it here, but it would be good if we could rely on the AggregationTestCase to verify all relevant aspects

**Pull request merged**